packages/push/src/instruction/exec/mod.rs

```diff
@@ -0,0 +1,94 @@
+mod dup;
+mod ifelse;
+mod unless;
+mod when;
+
+use self::{dup::Dup, ifelse::IfElse, unless::Unless, when::When};
+use super::{instruction_error::PushInstructionError, Instruction, NumOpens, PushInstruction};
+use crate::{
+    error::InstructionResult,
+    push_vm::{program::PushProgram, HasStack},
+};
+
+#[derive(Debug, strum_macros::Display, Clone, Eq, PartialEq)]
+pub enum ExecInstruction {
+    Noop, /* Maybe use `Noop(())` instead of `Noop(Noop)` when we get around to this. See
+           * 2024-03-31 stream chat for more. */
```

Comment:
Also has the advantage that, since we ended up using Noop in tests a lot, that () (or Noop for that matter), doesn't require any type bounds on the instruction impl and as such is usable as well when there isn't a bool stack or something.

Reply:
So I don't necessarily object to using `()` instead of a `Noop` type, but I'm also not complete sold on it either. My concerns:

- It seems very "clever", but I'm not yet clear on how it's _usefully_ clever.
- It definitely is inconsistent with how we handle other instructions.

If I was coming to this as someone not super experienced with Rust, for example, having `()` somewhere where I was expecting an instruction would probably be fairly surprising/confusing, and violate the Principle of Least Amazement.

So the question is whether using `()` would provide a clear advantage over a `Noop` type?

Reply:
So i was viewing this from a more long-term perspective where we might want to separate the push crate into potentially push-core, push-vm and push-instructions - in that case I would define the impl on () as an "example" inside the push-core crate (and as the only impl inside there) next to the trait. I wouldn't recommend external crates to use it tho, and the Exec::noop() instruction should be preferred over () most of the time, at least for users of our crate. Would also be consistent with n-tuples if we want to add them as a convenient method to type instruction groups, a empty tuple would then basically do nothing (or: a noop)

Reply:
So maybe the thing to do is just do `Exec::noop()` for now, and create an issue for tuples, bringing `()` in as the "base case" for the tuple types. The idea of `()` as an empty tuple makes a lot of sense to me and seems coherent/consistent in that context.

I'll create a `Noop` type and add that here, though, so the `Exec` instructions are all consistent. Then hopefully we can merge this in. 😮‍💨